### PR TITLE
Publish to Single-File

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
This change implements support for publishing apps to a [single file](https://github.com/dotnet/designs/blob/master/accepted/single-file/design.md).

* ``dotnet publish /p:PublishSingleFile=true`` causes the contents of the "original" publish directory to a single file in the actual publish directory
* Files marked with the meta-data ``<ExcludeFromSingleFile>false<ExcludeFromSingleFile>`` are left in the publish directory unbundled. This includes PDB files by default
* PDB files can be bundled into the single file by setting ``/p:IncludeSymbolsInSingleFile=true``

Publishing to a single file requires publishing wrt a RID using an apphost, because the generated file is the platform-specific AppHost executable with embedded dependencies.

### CODE REVIEW

**Comment by swaroop-sridhar** [comment]
@nguerrera @fadimounir  @sbomer  please review.
CC: @ericstj

**Comment by swaroop-sridhar** [comment]
@dotnet-bot  test this please

**Review by sbomer — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets [hunk 1]
```diff
@@ -85,6 +86,9 @@ Copyright (c) .NET Foundation. All rights reserved.
     -->
   <Target Name="PrepareForPublish">
 
+    <NETSdkError Condition="'$(_IsExecutable)' != 'true'"
```
[the remark this answers is not in the document]
Comment (sbomer) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: Oh you already fixed it. :)

[the remark this answers is not in the document]
Comment (swaroop-sridhar) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: Thanks @sbomer; as you know, I saw the feedback over your shoulder :)

**Comment by swaroop-sridhar** [comment]
Thanks @sbomer. Once the ILLinker change (https://github.com/dotnet/sdk/pull/3125) goes through, I'll add another test for Illink + R2R + Bundled app.

File: src/Tasks/Common/Resources/Strings.resx [hunk 2]
```diff
@@ -505,4 +505,16 @@ The following are names of parameters or literal values and should not be transl
...
+  <data name="CannotHaveSingleFileWithoutRuntimeIdentifier" xml:space="preserve">
+    <value>NETSDK1097: It is not supported to publish an application to a single-file without specifying a RuntimeIdentifier. Please either specify a RuntimeIdentifier or set PublishSingleFile to false.</value>
+    <comment>{StrBegin="NETSDK1097: "}</comment>
+  </data>  
+  <data name="CannotHaveSingleFileWithoutAppHost" xml:space="preserve">
+    <value>NETSDK1098: Applications published to a single-file are required to use the application host. Please either set PublishSingleFile to false or set UseAppHost to true.</value>
+    <comment>{StrBegin="NETSDK1098: "}</comment>
+  </data>
+  <data name="CannotHaveSingleFileWithoutExecutable" xml:space="preserve">
+    <value>NETSDK1099: Publishing to a single-file is currently only supported for executable applications.</value>
```
Comment (fadimounir) [inline src/Tasks/Common/Resources/Strings.resx]: >currently [](start = 54, length = 9)

I feel we should remove "currently". If we support this in the future, we can delete this error message

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets [hunk 3]
```diff
@@ -770,6 +792,47 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+  <Target Name="BundlePublishDirectory" 
+          Condition="'$(PublishSingleFile)' == 'true'"
+          DependsOnTargets="_ComputeFilesToBundle" 
+          Inputs="@(_FilesToBundle)"
+          Outputs="$(PublishDir)\$(AssemblyName)$(_NativeExecutableExtension)">
+
+    <PropertyGroup>
+      <!-- Handle the case where IncludeSymbolsInSingleFile may not be defined -->
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' == 'true'">true</EmbedPDBs>
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' != 'true'">false</EmbedPDBs>
```
Comment (fadimounir) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: We should also handle the .pdb/.map files created by crossgen here. We can do this later (I want to make some changes to the R2R related stuff first). Please file an issue to track this.

**Review by fadimounir — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets (hunk 3 above)
Comment (swaroop-sridhar) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: https://github.com/dotnet/core-setup/issues/5903

**Comment by swaroop-sridhar** [comment]
In this change https://github.com/dotnet/sdk/pull/3132/commits/1cc93522a00a961281242ded2e629a26158e6e83 @fadimounir I updated  a test case to use the new name `PublishReadyToRun` for triggering R2R compilation, now that https://github.com/dotnet/sdk/pull/3134 is checked in.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+        protected override void ExecuteCore()
+        {
+            Bundler bundler = new Bundler(AppHostName, OutputDir, EmbedPDBs, ShowDisgnosticOutput);
+            var fileSpec = new List<FileSpec>(FilesToBundle.Length);
+
+            foreach(var item in FilesToBundle)
+            {
+                fileSpec.Add(new FileSpec(sourcePath: item.ItemSpec,
+                                          bundleRelativePath: item.GetMetadata("RelativePath")));
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]:
```suggestion
                                          bundleRelativePath: item.GetMetadata(MetadataKeys.RelativePath)));
```

Not sure if I got that exactly right, but there's a class with constants for metadata. Use that.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        public string OutputDir { get; set; }
+        [Required]
+        public bool ShowDisgnosticOutput { get; set; }
+
+        protected override void ExecuteCore()
+        {
+            Bundler bundler = new Bundler(AppHostName, OutputDir, EmbedPDBs, ShowDisgnosticOutput);
+            var fileSpec = new List<FileSpec>(FilesToBundle.Length);
+
+            foreach(var item in FilesToBundle)
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]:
```suggestion
            foreach (var item in FilesToBundle)
```
nit: spacing

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets [hunk 6]
```diff
@@ -121,20 +127,28 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+    (except files that will be bundled, during PublishSingleFile).
     ============================================================
     -->
   <Target Name="_CopyResolvedFilesToPublishPreserveNewest"
           DependsOnTargets="_ComputeResolvedFilesToPublishTypes"
           Inputs="@(_ResolvedFileToPublishPreserveNewest)"
           Outputs="@(_ResolvedFileToPublishPreserveNewest->'$(PublishDir)%(RelativePath)')">
 
+    <ItemGroup>
+      <_ResolvedUnBundledFileToPublishPreserveNewest
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]:
```suggestion
      <_ResolvedUnbundledFileToPublishPreserveNewest 
```

nit on casing. Un is not a word on its own.

Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: (Obviously, rename all references to it too.)

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets [hunk 7]
```diff
@@ -719,6 +741,47 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+  <Target Name="BundlePublishDirectory" 
+          Condition="'$(PublishSingleFile)' == 'true'"
+          DependsOnTargets="_ComputeFilesToBundle" 
+          Inputs="@(_FilesToBundle)"
+          Outputs="$(PublishDir)\$(AssemblyName)$(_NativeExecutableExtension)">
+
+    <PropertyGroup>
+      <!-- Handle the case where IncludeSymbolsInSingleFile may not be defined -->
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' == 'true'">true</EmbedPDBs>
+      <EmbedPDBs Condition="'$(IncludeSymbolsInSingleFile)' != 'true'">false</EmbedPDBs>
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: EmbedPDBs is a pretty generic name since there is also a concept of embeddign PDB in dll (DebugType=Embedded). You can avoid having a new name, by defaulting IncludeSymbolsInSingleFile to false when empty. This is very idiomatic in msbuild, you will see it all over the place.

``` xml
<PropertyGroup>
   <IncludeSymbolsInSingleFile Condition="'$(IncludeSymbolsInSingleFile)' == ''">false</IncludeSymbolsInSingleFile>
</PropertyGroup>
```
I don't even think you need the comment if you do that, it is pretty obvious what this handles.

Then pass EmbedPdbs="$(IncludeSymbolsInSingleFile)

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+namespace Microsoft.NET.Build.Tasks
+{
+    public class GenerateBundle : TaskBase
+    {
+        [Required]
+        public ITaskItem[] FilesToBundle { get; set; }
+        [Required]
+        public string AppHostName { get; set; }
+        [Required]
+        public bool EmbedPDBs { get; set; }
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]:
```suggestion
        public bool IncludeSymbols { get; set; }
```

I think it is worth using the same terminology throughout, and the public facing name is symbols for this feature.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        [Required]
+        public string OutputDir { get; set; }
+        [Required]
+        public bool ShowDisgnosticOutput { get; set; }
+
+        protected override void ExecuteCore()
+        {
+            Bundler bundler = new Bundler(AppHostName, OutputDir, EmbedPDBs, ShowDisgnosticOutput);
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]:
```suggestion
            var bundler = new Bundler(AppHostName, OutputDir, EmbedPDBs, ShowDisgnosticOutput);
```

nit: type is obvious

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets [hunk 10]
```diff
@@ -719,6 +741,47 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+    </ItemGroup>
+
+  </Target>
+
+  <UsingTask TaskName="GenerateBundle" AssemblyFile="$(MicrosoftNETBuildTasksAssembly)" />
+  <Target Name="BundlePublishDirectory" 
+          Condition="'$(PublishSingleFile)' == 'true'"
+          DependsOnTargets="_ComputeFilesToBundle" 
+          Inputs="@(_FilesToBundle)"
+          Outputs="$(PublishDir)\$(AssemblyName)$(_NativeExecutableExtension)">
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]:
```suggestion
          Outputs="$(PublishDir)$(AssemblyName)$(_NativeExecutableExtension)">
```

We ensure trailing slash and you can see this is how PublishDir is used elsewhere. This is another common idiom in msbuild, it prevents accidental writing to root of disk if something is undefined. I once deleted my entire hard drive this way so I stick to this now :)

**Review by nguerrera — COMMENTED** [review]
Looks good overall. Some minor feedback.

**Comment by swaroop-sridhar** [comment]
@nguerrera: I've made the corrections you suggested in this commit: 9246cad 
Thanks.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+        public string OutputDir { get; set; }
+        [Required]
+        public bool ShowDisgnosticOutput { get; set; }
+
+        protected override void ExecuteCore()
+        {
+            var bundler = new Bundler(AppHostName, OutputDir, IncludeSymbols, ShowDisgnosticOutput);
+            var fileSpec = new List<FileSpec>(FilesToBundle.Length);
+
+            Array.ForEach(FilesToBundle,
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]: Why did you change this to Array.ForEach? I liked the foreach loop better.

Comment (swaroop-sridhar) [reply src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]: OK @nguerrera, I pushed 25c2b423608690a37da83dbdcecdc55138af67cb which revers back to the foreach() loop.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        [Required]
+        public string AppHostName { get; set; }
+        [Required]
+        public bool IncludeSymbols { get; set; }
+        [Required]
+        public string OutputDir { get; set; }
+        [Required]
+        public bool ShowDisgnosticOutput { get; set; }
+
+        protected override void ExecuteCore()
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]: I love how delightfully simple this task ended up being. :)

**Review by nguerrera — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+        [Required]
+        public ITaskItem[] FilesToBundle { get; set; }
+        [Required]
+        public string AppHostName { get; set; }
+        [Required]
+        public bool IncludeSymbols { get; set; }
+        [Required]
+        public string OutputDir { get; set; }
+        [Required]
+        public bool ShowDisgnosticOutput { get; set; }
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]:
```suggestion
        public bool ShowDiagnosticOutput { get; set; }
```

Just spotted a typo

Comment (swaroop-sridhar) [reply src/Tasks/Microsoft.NET.Build.Tasks/GenerateBundle.cs]: Thanks, fixed it in 77e4d367b557e788242cee0b3406b83b298b5b0a

**Comment by swaroop-sridhar** [comment]
@dotnet-bot  test dotnet-sdk-public-ci

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets (hunk 10 above)
Comment (ericstj) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: Funny story, a similar issue caused CoreFx build perf to tank once, the cause `$(SomeProperty)\**\*.extension`.  Someone renamed `SomeProperty` and CI passed...

**Review by ericstj — APPROVED** [review]
Nice!

**Comment by cup** [comment]
https://github.com/dotnet/cli/issues/7737
https://github.com/dotnet/core/issues/1278
https://github.com/dotnet/core/issues/2647
https://github.com/dotnet/coreclr/issues/17015
https://github.com/dotnet/coreclr/issues/20287
https://github.com/dotnet/corefx/issues/13329
https://github.com/dotnet/corert/issues/6819
https://github.com/dotnet/core-setup/pull/5286
https://github.com/dotnet/designs/pull/52
https://github.com/dotnet/sdk/pull/3074

https://github.com/dotnet/sdk/pull/3132
